### PR TITLE
make simplifaction rules agnostic to order of execution

### PR DESCRIPTION
**Pull request by normen662**

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/simplification/CompensateRecordConstructorRule.java [hunk 1]
```diff
@@ -59,28 +59,33 @@ public CompensateRecordConstructorRule() {
...
-            // i.e. the columns flowing results of values we care about.
-            //
-            for (final var childValueEntry : childResultPair.getRight().entrySet()) {
-                final var argumentValue = childValueEntry.getKey();
-                final var argumentValueCompensation = childValueEntry.getValue();
-                final var field = column.getField();
-                mergedMatchedValuesMap.putIfAbsent(argumentValue,
-                        new FieldValueCompensation(FieldValue.FieldPath.ofSingle(field.getFieldNameOptional().orElse(null), field.getFieldType(), i), argumentValueCompensation));
+                //
+                // No we have a column and the result we computed for all columns that do have results associated with them,
```
Comment (hatyo) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/simplification/CompensateRecordConstructorRule.java:76]:
```suggestion
                // Now we have a column and the result we computed for all columns that do have results associated with them,
```

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/simplification/CompensateRecordConstructorRule.java [hunk 2]
```diff
@@ -59,28 +59,33 @@ public CompensateRecordConstructorRule() {
...
 
-        for (int i = 0; i < recordConstructorValue.getColumns().size(); ++i) {
-            final var column = recordConstructorValue.getColumns().get(i);
-            final var childResultPair = call.getResult(column.getValue());
-            if (childResultPair == null) {
-                continue;
-            }
+        final var recordConstructorValueResult = call.getResult(recordConstructorValue);
+        if (recordConstructorValueResult != null) {
+            resultingMatchedValuesMap.put(recordConstructorValue, recordConstructorValueResult.getValue().get(recordConstructorValue));
```
Comment (hatyo) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/simplification/CompensateRecordConstructorRule.java:66]: Can we avoid the extra `Map` by doing something like:

```java
call.yield(recordConstructorValue, recordConstructorValueResult.getRight());
return;
```

Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/simplification/CompensateRecordConstructorRule.java:66]: There may be other values in that map that I don't want to return.

**Review by hatyo — APPROVED** [review]
LGTM.

**Pull request merged**